File: Math/Medium/LET_no2244_Minimum Rounds to Complete All Tasks.py
class Solution:
    def minimumRounds(self, tasks: List[int]) -> int:
        memo = collections.Counter(tasks)
        res = 0
        
        # If freq = 1, not possible, return -1
        # If freq = 2, needs one 2-tasks
        # If freq = 3, needs one 3-tasks
        # If freq = 3k, freq = 3 * k, needs k batchs.
        # If freq = 3k + 1, freq = 3 * (k - 1) + 2 + 2, needs k + 1 batchs.
        # If freq = 3k + 2, freq = 3 * k + 2, needs k + 1 batchs.
        
        for _, val in memo.items():
            if val < 2:
                return -1
            res += (val + 2) // 3
            
        return res

File: Math/Medium/LET_no2244_Minimum Rounds to Complete All Tasks.cpp
class Solution {
public:
    int minimumRounds(vector<int>& tasks) {
        unordered_map<int, int> memo;
        for (int num : tasks) {
            auto it = memo.find(num);
            if (it == memo.end()) 
                memo[num] = 1;
            else
                it->second++;
        }

        int res = 0;
        for (auto it = memo.begin(); it != memo.end(); it++) {
            int count = it->second;
            if (count < 2)
                return -1;
            res += (count + 2) / 3;
        }

        return res;
    }
};